wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,166 @@
+  private static final Set<String> subscribedMethods =
+      Collections.unmodifiableSet(new HashSet<String>() {
+        {
+          add("connect");
+          add("forceConnect");
+        }
+      });
+
+  public static final AwsWrapperProperty REGION_PROPERTY = new AwsWrapperProperty(
+      "customEndpointRegion", null,
+      "The region of the cluster's custom endpoints.");
+
+  public static final AwsWrapperProperty CUSTOM_ENDPOINT_INFO_REFRESH_RATE = new AwsWrapperProperty(
+      "customEndpointInfoRefreshRateMs", "30000",
+      "Controls how frequently custom endpoint monitors fetch custom endpoint info.");
+
+  static {
+    PropertyDefinition.registerPluginProperties(CustomEndpointPlugin.class);
+  }
+
+  protected final PluginService pluginService;
+  protected final RdsUtils rdsUtils = new RdsUtils();
+  protected final BiFunction<HostSpec, Region, RdsClient> rdsClientFunc;
+  protected CustomEndpointMonitor customEndpointMonitor;
+
+  public CustomEndpointPlugin(final PluginService pluginService, final Properties props) {
+    this(
+        pluginService,
+        (hostSpec, region) ->
+            RdsClient.builder()
+                .region(region)
+                .credentialsProvider(AwsCredentialsManager.getProvider(hostSpec, props))
+                .build());
+  }
+
+  public CustomEndpointPlugin(
+      final PluginService pluginService,
+      final BiFunction<HostSpec, Region, RdsClient> rdsClientFunc) {
+    this.pluginService = pluginService;
+    this.rdsClientFunc = rdsClientFunc;
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return subscribedMethods;
+  }
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, isInitialConnection, connectFunc);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, isInitialConnection, forceConnectFunc);
+  }
+
+  protected Connection connectInternal(
+      HostSpec hostSpec,
+      Properties props,
+      boolean isInitialConnection,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    final Connection conn = connectFunc.call();
+    if (!isInitialConnection || !this.rdsUtils.isRdsCustomClusterDns(hostSpec.getHost())) {
+      return conn;
+    }
+
+    // If we get here we are making an initial connection to a custom cluster URL
+    String customClusterHost = hostSpec.getHost();
+    CustomEndpointMonitor monitor = monitors.get(customClusterHost, MONITOR_EXPIRATION_NANO);
```

Comment:
monitors.computeIfPresent()?

Reply:
`monitors` is a SlidingExpirationCache, which does not expose a computeIfPresent method, but I just realized I can get rid of this block and clean things up, so we shouldn't need it